### PR TITLE
Support SSH urls

### PR DESCRIPTION
**Pull request by vinokurig**
### What does this PR do?
Extend each SCM provider implementation to support SSH urls.

### What issues does this PR fix or reference?

fixes https://github.com/eclipse/che/issues/22354

### How to test this PR?

1. Deploy Che with the PR image: `quay.io/eclipse/che-server:pr-539`
2. [Configure Che to use SSH keys for Git operations](https://github.com/devfile/devworkspace-operator/blob/main/docs/additional-configuration.adoc#configuring-devworkspaces-to-use-ssh-keys-for-git-operations)
3. Try to start a workspace from an SSH url of a **PUBLIC** repository to avoid configuring OAuth or PAT, see: https://github.com/eclipse-che/che-docs/pull/2639

See: 
* Dashboard shows a notification that the devfile has been fetched:
screenshot-eclipse-che apps cluster-c8wk5 c8wk5 sandbox2818 opentlc com-2023 08 08-16_06_34
* Workspace starts successfully

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [ ] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [ ] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [ ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

**Comment by ibuziuk** [comment]
@vinokurig could you please update the description e.g. ssh setup step and provide an image for testing

File: wsmaster/che-core-api-factory-azure-devops/src/main/java/org/eclipse/che/api/factory/server/azure/devops/AzureDevOpsUrl.java [hunk 1]
```diff
@@ -145,7 +146,17 @@ public String rawFileLocation(String fileName) {
   }
 
   public String getRepositoryLocation() {
-    return getRepoPathJoiner().add("_git").add(repository).toString();
+    if (isHTTPSUrl) {
+      return getRepoPathJoiner().add("_git").add(repository).toString();
+    }
+    return "git@ssh."
+        + hostName.substring(8)
```
Comment (tolusha) [inline wsmaster/che-core-api-factory-azure-devops/src/main/java/org/eclipse/che/api/factory/server/azure/devops/AzureDevOpsUrl.java:153]: I can see some magic numbers in the PR
Could you avoid that?

File: wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/scm/AuthorizingFileContentProvider.java [hunk 2]
```diff
@@ -80,7 +80,7 @@ private String fetchContent(
         String authorization;
         if (isNullOrEmpty(credentials)) {
           PersonalAccessToken token =
-              personalAccessTokenManager.get(remoteFactoryUrl.getHostName());
+              personalAccessTokenManager.get(remoteFactoryUrl.getProviderUrl());
```
Comment (tolusha) [inline wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/scm/AuthorizingFileContentProvider.java:83]: Will it affect existed PATs?

**Review by tolusha — COMMENTED** [review]
Tested for Azure DevOps

File: wsmaster/che-core-api-factory-azure-devops/src/main/java/org/eclipse/che/api/factory/server/azure/devops/AzureDevOpsUrl.java (hunk 1 above)
Comment (vinokurig) [reply wsmaster/che-core-api-factory-azure-devops/src/main/java/org/eclipse/che/api/factory/server/azure/devops/AzureDevOpsUrl.java]: done

File: wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/scm/AuthorizingFileContentProvider.java (hunk 2 above)
Comment (vinokurig) [reply wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/scm/AuthorizingFileContentProvider.java:83]: It won't. This is just a refactoring of a method which retrieves SCM endpoint. The output is the same but the method is different.

**Comment by tolusha** [comment]
```bash
test-project (main) $ git remote -v
origin  git@ssh.dev.azure.com:v3/my-abazko/test-project/test-project (fetch)
origin  git@ssh.dev.azure.com:v3/my-abazko/test-project/test-project (push)
test-project (main) $ touch test-file
test-project (main) $ git add -A
test-project (main) $ git commit -m "Add file" --signoff 
[main 3205e4c] Add file
 1 file changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 test-file
test-project (main) $ git push origin 
Warning: Permanently added 'ssh.dev.azure.com,40.74.28.27' (RSA) to the list of known hosts.
Enumerating objects: 4, done.
Counting objects: 100% (4/4), done.
Delta compression using up to 4 threads
Compressing objects: 100% (2/2), done.
Writing objects: 100% (3/3), 328 bytes | 109.00 KiB/s, done.
Total 3 (delta 0), reused 0 (delta 0), pack-reused 0
remote: Analyzing objects... (3/3) (28 ms)
remote: Storing packfile... done (63 ms)
remote: Storing index... done (60 ms)
To ssh.dev.azure.com:v3/my-abazko/test-project/test-project
   dec2ba5..3205e4c  main -> main
test-project (main) $ 
```

**Comment by ibuziuk** [comment]
@ScrewTSW when you will be doing review please verify against bitbucket

**Comment by vinokurig** [comment]
@ibuziuk 
> Also, wondering if we can make the error more explicit

The error is thrown because Che could not find GitHub oauth provider in its list of registered providers. The same error appears when a private https url is used without Oauth and PAT being configured. This question worth a separate discussion.

**Review by ibuziuk — APPROVED** [review]

**Comment by ScrewTSW** [comment]
LGTM, verified using the prcheck image - AOK

**Review by ScrewTSW — APPROVED** [review]

**Comment by l0rd** [comment]
After this SSH URLs other than GitHub/GitLab/BB/Azure don't work at all. It doesn't matter if the repo is public or private, the workspace will fail to start. This is a regression from what we had before because the workspace could start although the devfile was ignored. (I have tried with sourcehut for instance and I verified that it fails at the step "Inspecting repo for a devfile" with message "Cannot build factory with any of the provided parameters. Please check parameters correctness, and resend query."). 

Moreover, as mentioned by @ibuziuk above, for GitHub/GitLab/BB/Azure SSH URLs it still requires a PAT secret (beyond the SSH private key) and there is no obvious value using the SSH URL. But I may be wrong here because using SSH key to pull/push rather then PAT can avoid problems with token expiration (to verify) so that, at least, have some value for users.

**Comment by vinokurig** [comment]
@l0rd 
> After this SSH URLs other than GitHub/GitLab/BB/Azure don't work at all. It doesn't matter if the repo is public or private, the workspace will fail to start. This is a regression from what we had before because the workspace could start although the devfile was ignored. (I have tried with sourcehut for instance and I verified that it fails at the step "Inspecting repo for a devfile" with message "Cannot build factory with any of the provided parameters. Please check parameters correctness, and resend query.")

filed a bug report: https://github.com/eclipse/che/issues/22473

**Pull request merged**